src/main/scala/util/AsyncQueue.scala

```diff
@@ -61,9 +72,16 @@ class AsyncQueueSink[T <: Data](gen: T, depth: Int, sync: Int) extends Module {
     val ridx = UInt(OUTPUT, width = bits+1)
     val widx = UInt(INPUT,  width = bits+1)
     val mem  = Vec(depth, gen).asInput
+    // Reset for the other side
+    val source_reset_n = Bool().flip
   }
 
-  val ridx = GrayCounter(bits+1, io.deq.fire())
+  // extend the source reset to a full cycle (assertion latency <= 1 cycle)
+  val catch_reset_n = AsyncResetReg(Bool(true), clock, !io.source_reset_n)
+  // reset_n has a 1 cycle shorter path to valid than widx does
+  val reset_n = AsyncGrayCounter(catch_reset_n.asUInt, sync)(0)
```

Comment:
Every implementation of an AsyncQueue assumes that independent 1-bit signals sent from one side will arrive no more than one cycle apart in the target side. It's the whole reason why Gray codes are a safe way to transport a counter... So, yes, I am assuming that reset_n will not arrive more than 1 cycle later than widx. Then, I ensure the path to valid/ready is one cycle shorter, guaranteeing that there can be no spurious valid/ready.